### PR TITLE
[rollout] feat: support multi-batch inference in rollout

### PR DESCRIPTION
**Pull request by wsc16**
### What does this PR do?
Support multi-batch inference in rollout based on the [async batch inference feature](https://github.com/vllm-project/vllm-omni/pull/2729).

### Checklist Before Starting

- [ ] Search for similar PRs. Paste at least one query link here: ...
- [x] Format the PR title as `[{modules}] {type}: {description}` (This will be checked by the CI)
  - `{modules}` include `fsdp`, `vllm_omni`, `rollout`, `trainer`, `ci`, `training_utils`, `recipe`, `ray`, `worker`, `single_controller`, `misc`, `perf`, `model`, `algo`, `env`, `tool`, `ckpt`, `doc`, `data`, `cfg`, `reward`, `diffusion`, `omni`, `tests`, `docker`
  - If this PR involves multiple modules, separate them with `,` like `[diffusion, doc]`
  - `{type}` is in `feat`, `fix`, `refactor`, `chore`, `test`
  - If this PR breaks any API (CLI arguments, config, function signature, etc.), add `[BREAKING]` to the beginning of the title.
  - Example: `[BREAKING][diffusion, fsdp] feat: new rollout scheduler`

### Test
The full-parameter training of the QwenImage model was verified on Atlas 800T A3 (CANN 9.0.0) with 16 cards. The performance during the rollout phase improved from 220s to 60s. And the training results are as follow:

actor
image

reward
image

val
image

### API and Usage Example
Add the following configuration to the training script to enable stepwise execution:
- `actor_rollout_ref.rollout.max_num_seqs=16`
- `+actor_rollout_ref.rollout.engine_kwargs.vllm_omni.step_execution=True`

```bash
bash examples/flowgrpo_trainer/run_qwen_image_ocr_npu.sh 
```

### Design & Code Changes

> Demonstrate the high-level design if this PR is complex, and list the specific changes.

### Checklist Before Submitting

> [!IMPORTANT]
> Please check all the following items before requesting a review, otherwise the reviewer might deprioritize this PR for review.

- [ ] Read the [Contribute Guide](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md).
- [ ] Apply [pre-commit checks](https://github.com/verl-project/verl-omni/blob/main/CONTRIBUTING.md#code-linting-and-formatting): `pre-commit install && pre-commit run --all-files --show-diff-on-failure --color=always`
- [ ] Add / Update the documentation.
- [ ] Add unit or end-to-end test(s) to [the CI workflow](https://github.com/verl-project/verl-omni/tree/main/.github/workflows) to cover all the code. If not feasible, explain why: ...

### CODE REVIEW

**Comment by SamitHuang** [comment]
related PR: 
https://github.com/verl-project/verl-omni/pull/81   step-wise batching adaptation
https://github.com/vllm-project/vllm-omni/pull/4070   sample-wise batching support in vllm-omni

Can you try whether https://github.com/verl-project/verl-omni/pull/81  and verify whether step-wise batching reduce the same amount of rollout time? 

I think we need a RFC design for this rollout batching. Suppose at the rollout phase, 512 inference requests arrives at vllm-omni engine. The vllm-omni engine should be able to do
1) request-level batching, group requests into different batches, where the batch size can be controlled max-concurrecy. This is enough for flow-grpo on-policy RL, since 512 rollout requests arrive at the same time.
2) step-level batching, group request at the level of diffuse forward steps. This is more general. Can be more efficient for async RL.

@knlnguyen1802  @bjf-frz

**Comment by zhtmike** [comment]
Agree with @SamitHuang, Step batching should cover the usage for async batching. And it should be more flexible, fully utillize the async agent loop / async rewards feature

**Comment by wsc16** [comment]
> related PR: #81 step-wise batching adaptation [vllm-project/vllm-omni#4070](https://github.com/vllm-project/vllm-omni/pull/4070) sample-wise batching support in vllm-omni
> 
> Can you try whether #81 and verify whether step-wise batching reduce the same amount of rollout time?
> 
> I think we need a RFC design for this rollout batching. Suppose at the rollout phase, 512 inference requests arrives at vllm-omni engine. The vllm-omni engine should be able to do
> 
> 1. request-level batching, group requests into different batches, where the batch size can be controlled max-concurrecy. This is enough for flow-grpo on-policy RL, since 512 rollout requests arrive at the same time.
> 2. step-level batching, group request at the level of diffuse forward steps. This is more general. Can be more efficient for async RL.
> 
> @knlnguyen1802 @bjf-frz

Both are adaptations for the stepwise execution. It seems to be working fine and I will test with the PR: #81.

**Comment by wsc16** [comment]
> > related PR: #81 step-wise batching adaptation [vllm-project/vllm-omni#4070](https://github.com/vllm-project/vllm-omni/pull/4070) sample-wise batching support in vllm-omni
> > Can you try whether #81 and verify whether step-wise batching reduce the same amount of rollout time?
> > I think we need a RFC design for this rollout batching. Suppose at the rollout phase, 512 inference requests arrives at vllm-omni engine. The vllm-omni engine should be able to do
> > 
> > 1. request-level batching, group requests into different batches, where the batch size can be controlled max-concurrecy. This is enough for flow-grpo on-policy RL, since 512 rollout requests arrive at the same time.
> > 2. step-level batching, group request at the level of diffuse forward steps. This is more general. Can be more efficient for async RL.
> > 
> > @SamitHuang @zhtmike  @knlnguyen1802 @bjf-frz 
> 
> Both are adaptations for the stepwise execution. It seems to be working fine and I will test with the PR: #81.

I test with PR #81, it has the same optimization effect. While in the scenario of mixed-precision training, there is a problem of inconsistent data types that needs to be fixed.
```diff
diff --git a/verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py b/verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py
index ab34229..02e0d78 100644
--- a/verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py
+++ b/verl_omni/pipelines/qwen_image_flow_grpo/vllm_omni_rollout_adapter.py
@@ -261,12 +261,13 @@ class QwenImagePipelineWithLogProb(QwenImagePipeline):
             negative_prompt_embeds_mask = None
 
         num_channels_latents = self.transformer.in_channels // 4
+        # Use float32 for latents to match scheduler requirements
         latents = self.prepare_latents(
             batch_size * num_images_per_prompt,
             num_channels_latents,
             height,
             width,
-            prompt_embeds.dtype,
+            torch.float32,
             self.device,
             generator,
             None,
@@ -453,7 +454,7 @@ class QwenImagePipelineWithLogProb(QwenImagePipeline):
             latents, log_prob, _, _ = self.scheduler.step(
                 noise_pred.float(),
                 timestep_value,
-                latents,
+                latents.float(),
                 generator=generator,
                 noise_level=cur_noise_level,
                 sde_type=sde_type,
@@ -507,9 +508,9 @@ class QwenImagePipelineWithLogProb(QwenImagePipeline):
             cur_noise_level = 0.0
 
         new_latents, log_prob, _, _ = state.scheduler.step(
-            noise_pred,
+            noise_pred.to(torch.float32),
             timestep_value,
-            state.latents,
+            state.latents.to(torch.float32),
             generator=state.sampling.generator,
             noise_level=cur_noise_level,
             sde_type=state.sde_type,
```
